sql/sql_parse.cc

```diff
@@ -7835,6 +7835,39 @@ static uint kill_one_thread(THD *thd, my_thread_id id, bool only_kill_query)
 }
 
 #ifdef WITH_WSREP
+static bool wsrep_should_retry_in_autocommit(enum_sql_command &sql_command)
+{
+  /*
+    We are here could mean that the query resulted in a cluster-wide
+    conflict and had to be aborted. While it happened, it is possible that
+    the client may have already received partial data from server and may
+    have been waiting for the OK/EOF packet (usually sent by
+    THD::send_statement_status() in dispatch_command()) to report it to the
+    user.
+
+    When retry is performed in such a case, the server shall start sending
+    result and field metadata once again and this would cause the client
+    program to receive unexpected metadata information in place of an
+    OK/EOF packet and thus causes the client to error out with Malformed
+    packet error.
+
+    So, we avoid retries for such queries that return result set to client,
+    but cannot be run in TOI and can be killed by a TOI.
+
+    As of now, we only do this check for CHECK TABLE and SELECT, and if the
+    same symptom is found for other commands, then please add it to the
+    below list.
+  */
+  switch (sql_command)
```

Comment:
The ticket mentions also problems with ANALYZE TABLE. Maybe worthy adding here as well?

Reply:
Even though ticket mentions about ANALYZE TABLE, I was never able to see the crash with ANALYZE TABLE. This is because ANALYZE/OPTIMIZE TABLE are replicated as TOI and can never get bf aborted by applier threads (as the whole cluster will be isolated), but problem was seen only with CHECK TABLE since is not replicated to the cluster. So, it is not required here.